data/src/main/java/com/microsoft/azure/kusto/data/AadAuthenticationHelper.java

```diff
@@ -3,300 +3,282 @@
-        AuthenticationContext context;
-        AuthenticationResult result;
-        ExecutorService service = null;
-        try {
-            service = Executors.newSingleThreadExecutor();
-            context = new AuthenticationContext(aadAuthorityUri, true, service);
-
-            Future<AuthenticationResult> future = context.acquireToken(
-                    clusterUrl, CLIENT_ID, userUsername, userPassword,
-                    null);
-            result = future.get();
-        } catch (InterruptedException | ExecutionException | MalformedURLException e) {
-            throw new DataClientException(clusterUrl, "Error in acquiring UserAccessToken", e);
-        } finally {
-            if (service != null) {
-                service.shutdown();
-            }
+    protected IAuthenticationResult acquireNewAccessToken() throws DataServiceException, DataClientException {
+        switch (authenticationType) {
+            case AAD_APPLICATION_KEY:
+                return acquireWithAadApplicationKey();
+            case AAD_APPLICATION_CERTIFICATE:
+                return acquireWithAadApplicationClientCertificate();
+            case AAD_USER_PROMPT:
+                return acquireWithUserPrompt();
+            default:
+                throw new DataClientException("Authentication type '" + authenticationType.name() + "' is invalid");
         }
+    }
 
-        if (result == null) {
-            throw new DataServiceException(clusterUrl, "acquireAadUserAccessToken got 'null' authentication result");
-        }
-        return result;
+    private boolean isInvalidToken() {
+        return lastAuthenticationResult == null || lastAuthenticationResult.expiresOnDate().before(dateInAMinute());
     }
 
-    private AuthenticationResult acquireAadApplicationAccessToken() throws DataServiceException, DataClientException {
-        AuthenticationContext context;
-        AuthenticationResult result;
-        ExecutorService service = null;
+    protected IAuthenticationResult acquireWithAadApplicationKey() throws DataServiceException, DataClientException {
+        IAuthenticationResult result;
         try {
-            service = Executors.newSingleThreadExecutor();
-            context = new AuthenticationContext(aadAuthorityUri, true, service);
-            Future<AuthenticationResult> future = context.acquireToken(clusterUrl, clientCredential, null);
-            result = future.get();
-        } catch (InterruptedException | ExecutionException | MalformedURLException e) {
-            throw new DataClientException(clusterUrl, "Error in acquiring ApplicationAccessToken", e);
-        } finally {
-            if (service != null) {
-                service.shutdown();
-            }
+            lastConfidentialClientApplication = ConfidentialClientApplication.builder(clientId, clientSecret).authority(aadAuthorityUrl).build();
```

Comment:
It looked that way to me as well, but a ClientApplication object that had acquired a token will be different than a new ClientApplication object due to its cache. And we branch based on this cache later in the code.

Reply:
I don't follow.
You rely on this for acquire token silent, why stop relying on it if it has failed, i.e. if the cache entry is missing or invalid.

Reply:
I was concerned that it might fail to obtain the refresh token, where creating a new ClientApplication object would resolve the issue. I think you're right though that that's not a legitimate concern.

I do still see an extremely rare benefit to creating new ClientApplication objects (user changed their account's username, assuming that's possible), but I concede that that's unlikely to ever actually make a difference.

I'll build it once as you suggest.

Reply:
Sorry to retract my latest reply, but note that this in practice should only create the object once. Each of these acquireNewToken methods are only called once (unless getting the refresh token fails, which should hardly ever happen if ever, and anyway I'm taking out any further calls to acquire a new token per your request because of this rarity/impossibility of being necessary).